---
title: "Frankel TCDD scRNA-seq analysis"
output: html_document
date: "2023-12-13"
---

# Libraries

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#library(devtools)
library(tidyverse)
library(readxl)
library(ggrepel)
library(writexl)

library(Seurat)
library(presto) # https://github.com/immunogenomics/presto
library(escape)
library(babelgene)

```

## Setup

```{r setup, include=FALSE}

code_dir <- file.path("/mctp/share/users/mkiyer/proj/frankel_smoking")
working_dir <- file.path("/mctp/share/users/mkiyer/analysis/frankel_smoking")
plot_dir <- file.path(working_dir, "plots")

ahr_signature_xlsx_file <- file.path(code_dir, "ahr_signature_opitz_cell_2020.xlsx")
seurat_object_file <- file.path(working_dir, "OrthotopicTCDD.RData")

for (d in c(working_dir, plot_dir)) {
  if (!dir.exists(d)) {
    dir.create(d)
  }
}

```

## read published AHR gene signature and convert to mouse

```{r ahr signature read, include=FALSE}

# AHR signature (human gene symbols)
ahr_opitz_cell_human_genes <- read_xlsx(ahr_signature_xlsx_file)
ahr_opitz_cell_human_genes <- ahr_opitz_cell_human_genes %>% pull(Gene)

# convert to mouse genes
x <- orthologs(ahr_opitz_cell_human_genes, species="mouse", human=TRUE)
ahr_opitz_cell_mouse_genes <- x %>% pull(symbol)

# literature curated ahr signaling genes
ahr_literature_genes <- c("Ahr", "Cyp1a1", "Cyp1b1", "Tiparp", "Ahrr", "Thbs1")

```

## Load single cell object

```{r load Rdata}

# Seurat
load(seurat_object_file)

```

## Dimplots and feature plots

```{r umap}

# reference umap
p <- DimPlot(object = OrthotopicTCDD, 
             cols = pals::cols25(),
             reduction = "umap", 
             label = TRUE, 
             pt.size = 1)
ggsave(file.path(plot_dir, "umap.pdf"), p, width=10, height=10)

# umap by condition
p <- DimPlot(object = OrthotopicTCDD, 
             cols = pals::cols25(),
             reduction = "umap", 
             label = FALSE, 
             pt.size = 1, 
             split.by = 'Group')
ggsave(file.path(plot_dir, "umap_by_group.pdf"), p, width=20, height=10)

# feature plots of subset of ahr literature genes
genes <- c("Cyp1a1", "Cyp1b1", "Tiparp", "Ahrr")
p <- FeaturePlot(OrthotopicTCDD, 
                 features=genes, 
                 alpha=0.8,
                 order=TRUE,
                 reduction="umap",
                 split.by="Group")
f <- file.path(plot_dir, "feature_plot_ahr_genes.pdf")
ggsave(f, p, width=12, height=20)


# T cells alone
# x <- subset(x = OrthotopicTCDD, 
#             idents = c("T cells", "Proliferating T cells", "NK cells", "Myeloid & T cell mix"))
# p <- FeaturePlot(x, 
#                  features=genes, 
#                  alpha=0.8,
#                  order=TRUE,
#                  reduction="umap",
#                  split.by="Group")
# f <- file.path(plot_dir, "test.pdf")
# ggsave(f, p, width=12, height=20)

```

## Differential expression between vehicle and TCDD

```{r differential expression}

de <- NULL
for (myident in levels(OrthotopicTCDD)) {
  print(myident)
  x <- FindMarkers(OrthotopicTCDD, 
                   logfc.threshold=0.1,
                   test.use="wilcox",
                   min_pct=0.01,
                   verbose=TRUE,
                   ident.1="TCDD",
                   ident.2="Vehicle",
                   group.by="Group",
                   subset.ident = myident)
  x$ident <- myident
  x <- as_tibble(x, rownames="gene")
  de <- bind_rows(de, x)
}

# perform all TCDD vs vehicle
print("all cells")
x <- FindMarkers(OrthotopicTCDD, 
                 logfc.threshold=0.1,
                 test.use="wilcox",
                 min_pct=0.01,
                 verbose=TRUE,
                 ident.1="TCDD",
                 ident.2="Vehicle",
                 group.by="Group")
x$ident <- "All_cells"
x <- as_tibble(x, rownames="gene")
de <- bind_rows(de, x)

# write de results
de_filtered <- filter(de, abs(avg_log2FC) > 1, p_val_adj < 1e-5)
write_xlsx(de, path=file.path(working_dir, "de_results.xlsx"))

# add difference in pct detection
# de <- de %>% mutate(pct_diff = pct.1 - pct.2)



```

## plot DE genes

```{r de gene plots}

plot_de_volcano <- function(res, nlabel=25, 
                            highlight_genes=NULL, 
                            label_genes=NULL,
                            padj_threshold=1e-10,
                            log2fc_threshold=1,
                            log_base=10) {
  # label de genes
  x <- res %>% mutate(
    de = case_when(p_val_adj > padj_threshold ~ "no",
                   avg_log2FC > log2fc_threshold ~ "up",
                   avg_log2FC < -log2fc_threshold ~ "dn",
                   TRUE ~ "no")
  )
  
  # select the top n up/down genes that meet thresholds for DE for labeling
  de_up <- filter(x, de == "up") %>% slice_max(avg_log2FC, n=nlabel)
  de_dn <- filter(x, de == "dn") %>% slice_min(avg_log2FC, n=nlabel)
  # label highlighted genes that are also DE
  highlight_data <- filter(x, gene %in% highlight_genes)
  highlight_de_up <- filter(highlight_data, de=="up")
  highlight_de_dn <- filter(highlight_data, de=="dn")
  highlight_data <- distinct(highlight_data)
  # start with specifically labeled genes
  label_data <- filter(x, gene %in% label_genes)
  # add on top up/dn de as well as highlighted de genes
  label_data <- bind_rows(label_data, de_up, de_dn, highlight_de_up, highlight_de_dn)
  label_data <- distinct(label_data)

  p <- ggplot(x, aes(x=avg_log2FC, y=-log(p_val_adj+1e-300, log_base), color=de)) +
    geom_point(size=1, alpha=0.7) +
    geom_point(data=highlight_data, size=2, alpha=1, color="purple", pch=21, stroke=2) +
    geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +
    scale_color_manual(values=c("no"="grey", "dn"="blue", "up"="red")) +
    scale_size_manual(values=c("no"=1, "dn"=2, "up"=2)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(axis.line = element_line(color = "black")) +
    labs(x="avg_log2FC", y=paste0("-log(padj,", log_base, ")"))
  return(p)
}


for (myident in unique(de$ident)) {
  x <- dplyr::filter(de, ident == myident)
  p <- plot_de_volcano(x, label_genes = ahr_literature_genes,
                       highlight_genes = ahr_opitz_cell_mouse_genes) + 
    labs(title="Volcano Plot", subtitle=myident)
  ident_filename <- gsub(" ", "_", myident)
  f <- paste0("volcano_de_", ident_filename, ".pdf")
  ggsave(file.path(plot_dir, f), p, width=8, height=8)
  f <- paste0("volcano_de_", ident_filename, ".png")
  ggsave(file.path(plot_dir, f), p, width=8, height=8)
}


```


## AHR signature score enrichment

```{r enrichment of ahr signature}
# define gene set
gs <- list(ahr_signature = ahr_opitz_cell_mouse_genes)

# run enrichment
es <- enrichIt(obj = OrthotopicTCDD, 
               gene.sets = gs, 
               groups = 1000, 
               cores = 4, 
               min.size = 5)

# add enrichment results as metadata to Seurat object
OrthotopicTCDD <- Seurat::AddMetaData(OrthotopicTCDD, es)

```

## Plot AHR signature

```{r ahr plot}

# Cyp1a1, Cyp1b1, 
# Tiparp (TCDD-inducible poly adp-ribose polymerase)
# Ahrr (Ahr repressor), 
# Thbs1 (thrombospondin 1

genes <- c("Ahr", "Cyp1a1", "Cyp1b1", "Tiparp", "Ahrr", "Thbs1")

for (g in genes) {
  p <- VlnPlot(OrthotopicTCDD, features = c(g), split.by="Group", group.by = "collapsed_clusters", pt.size = 0.1, combine = FALSE)
  p <- patchwork::wrap_plots(plots = p, ncol = 1)
  f <- file.path(plot_dir, paste0("violinplot_by_cell_type_", g))
  ggsave(paste0(f, ".pdf"), p, width=10, height=5)
  ggsave(paste0(f, ".png"), p, width=10, height=5)
}

# AHR signature (~160 genes)
p <- VlnPlot(OrthotopicTCDD, features = c("ahr_signature"), split.by="Group", group.by = "collapsed_clusters", pt.size = 0.1, combine = FALSE)
p <- patchwork::wrap_plots(plots = p, ncol = 1)
f <- file.path(plot_dir, paste0("violinplot_by_cell_type_ahr_opitz_signature"))
ggsave(paste0(f, ".pdf"), p, width=10, height=5)
ggsave(paste0(f, ".png"), p, width=10, height=5)

p <- VlnPlot(OrthotopicTCDD, features = c("Ahr"), split.by="Group", group.by = "collapsed_clusters", pt.size = 0.1, combine = FALSE)
p <- patchwork::wrap_plots(plots = p, ncol = 1)
f <- file.path(plot_dir, "ahr")
ggsave(paste0(f, ".pdf"), p, width=10, height=5)
ggsave(paste0(f, ".png"), p, width=10, height=5)


p <- VlnPlot(OrthotopicTCDD, features = c("Cyp1a1", "Cyp1b1", "Tiparp", "Ahrr", "Thbs1"), split.by="Group", group.by = "collapsed_clusters", pt.size = 0.1, combine = FALSE)
p <- patchwork::wrap_plots(plots = p, ncol = 1)
f <- file.path(plot_dir, "ahr_activation_genes")
ggsave(paste0(f, ".pdf"), p, width=12, height=20)
ggsave(paste0(f, ".png"), p, width=12, height=20)

# FetchData(object = OrthotopicTCDD, vars = c('ahr_signature'))

```


## External code provided

```{r}

#Sample Info
#BG-1:Vehicle 1
#BG-2:Vehicle 2
#BG-3:TCDD 1
#BG-4:TCDD 2
#Orthotopic Tumor Cell line-7940B

#Load Dataset, make sure that genes are renamed to features
# Vehicle1.data <- Read10X("~/Desktop/TCDD Orthotopic scRNA-seq/BG-1") 
# Vehicle2.data <- Read10X("~/Desktop/TCDD Orthotopic scRNA-seq/BG-2") 
# TCDD1.data <- Read10X("~/Desktop/TCDD Orthotopic scRNA-seq/BG-3") 
# TCDD2.data <- Read10X("~/Desktop/TCDD Orthotopic scRNA-seq/BG-4") 


#Create Seurat objects
Vehicle1  <- CreateSeuratObject(counts = Vehicle1.data, project = 'Vehicle1', min.cells = 3, min.features = 100)
Vehicle2 <- CreateSeuratObject(counts = Vehicle2.data, project = 'Vehicle2', min.cells = 3, min.features = 100)
TCDD1 <- CreateSeuratObject(counts = TCDD1.data, project = 'TCDD1', min.cells = 3, min.features = 100)
TCDD2 <- CreateSeuratObject(counts = TCDD2.data, project = 'TCDD2', min.cells = 3, min.features = 100)

#Group MetaData
Vehicle1$Group <-"Vehicle"
Vehicle2$Group <-"Vehicle"
TCDD1$Group<-"TCDD"
TCDD2$Group <-"TCDD"
#Sample MetaData
Vehicle1$Sample <-"Vehicle 1"
Vehicle2$Sample <-"Vehicle 2"
TCDD1$Sample<-"TCDD 1"
TCDD2$Sample <-"TCDD 2"

#Merge samples
OrthotopicTCDD <- merge(Vehicle1, y =c(Vehicle2, TCDD1, TCDD2), 
                             add.cell.ids = c("Vehicle1", "Vehicle2","TCDD1","TCDD2"))
levels(OrthotopicTCDD)

# Changing between meta.data for identities- you can change this by altering what you input into your metadata
Idents(object = OrthotopicTCDD) <- 'Group'
#Check active Identity
levels(OrthotopicTCDD)

#Percent Mito
OrthotopicTCDD[["percent.mt"]] <- PercentageFeatureSet(OrthotopicTCDD, pattern = "^mt-")
VlnPlot(OrthotopicTCDD, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = .01)

Idents(OrthotopicTCDD) <- "Group"
table(OrthotopicTCDD@active.ident)

#Vehicle   #TCDD
#26331    #27717

#Filter genes further
OrthotopicTCDD <- subset(OrthotopicTCDD, subset = nCount_RNA > 1200 & nCount_RNA < 100000 & percent.mt < 15)
VlnPlot(OrthotopicTCDD, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = .01)

Idents(OrthotopicTCDD) <- "Group"
table(OrthotopicTCDD@active.ident)

#Vehicle    TCDD 
#  20995   20922 


#Normalization ScaleData and UMAP Generation
OrthotopicTCDD <- NormalizeData(OrthotopicTCDD)
OrthotopicTCDD <- FindVariableFeatures(OrthotopicTCDD, selection.method = "vst", nfeatures = 2000)
OrthotopicTCDD <- ScaleData(OrthotopicTCDD, verbose = T, features = row.names(OrthotopicTCDD))
OrthotopicTCDD <- RunPCA(object = OrthotopicTCDD)
stdev <- OrthotopicTCDD@reductions$pca@stdev
var <- stdev^2
sum(var[1:30])/ sum(var) #include a number that leads to 0.90 var

PCNum = 30 #put the dimensions here

#Find Neighbors + Find Clusters (without harmony batch correction)

OrthotopicTCDD <- FindNeighbors(object = OrthotopicTCDD, dims = 1:PCNum)
OrthotopicTCDD <- FindClusters(object = OrthotopicTCDD, resolution = 1.8)

#Run UMAP and get unlabelled cluster UMAP and violin plot
OrthotopicTCDD<- RunUMAP(object = OrthotopicTCDD, dims = 1:PCNum)
DimPlot(object = OrthotopicTCDD, reduction = "umap", label = FALSE, pt.size = 1, split.by = 'Group')

save(OrthotopicTCDD,file="OrthotopicTCDD.RData")
FeaturePlot(object = OrthotopicTCDD, features = c("Cd3e","Cd8a","Cd4","Nkg7"), cols = c("grey", "red"), reduction = "umap", pt.size = .5)# T cells 
FeaturePlot(object = OrthotopicTCDD, features = c("Mki67"), cols = c("grey", "red"), reduction = "umap", pt.size = .5)#proliferating Cells 
FeaturePlot(object = OrthotopicTCDD, features = c("Col1a2","Pdgfra","Pdgfrb"), cols = c("grey", "red"), reduction = "umap", pt.size = .5)#Fibroblast
FeaturePlot(object = OrthotopicTCDD, features = c("Krt19", "Muc1", "Krt18","Cdh1","Msln"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Epithelial
FeaturePlot(object = OrthotopicTCDD, features = c("Epcam","Krt19", "Muc1", "Krt18","Cdh1","Msln", "Spink1", "Amy1","Prss1"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Epithelial
FeaturePlot(object = OrthotopicTCDD, features = c("S100a9","S100a8","Mmp9","Il1r2"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Neutrophils
FeaturePlot(object = OrthotopicTCDD, features = c("Vwf", "Cdh5",'Erg'), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Endothelial
FeaturePlot(object = OrthotopicTCDD, features = c("Ms4a1", "Cd19", "Cd79a"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #B cells
FeaturePlot(object = OrthotopicTCDD, features = c("C1qb","C1qa","C1qc","Lyz2"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #TAMs
FeaturePlot(object = OrthotopicTCDD, features = c("Try4","Amy2a2"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Acinar 
FeaturePlot(object = OrthotopicTCDD, features = c("Itgae","Xcr1","Clec9a","Cd209a"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Dcs
FeaturePlot(object = OrthotopicTCDD, features = c("Itgam"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Dcs
FeaturePlot(object = OrthotopicTCDD, features = c("Ptprc"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #Immune Cells

#Lable Clusters
Idents(OrthotopicTCDD) <- "Group"
new_order <- c("Vehicle", "TCDD")
OrthotopicTCDD@active.ident <- factor(OrthotopicTCDD@active.ident, levels = new_order)
OrthotopicTCDD[["Group"]] <- OrthotopicTCDD@active.ident 
Idents(OrthotopicTCDD) <- "seurat_clusters"
OrthotopicTCDD <- RenameIdents(OrthotopicTCDD,
                          "0" = "TAMs", 
                          "1" = "TAMs", 
                          "2" = "Fibroblast 1", 
                          "3" = "Fibroblast 1", 
                          "4" = "Neutrophils", 
                          "5" = "TAMs", 
                          "6" = "TAMs", 
                          "7" = "Epithelial", 
                          "8" = "TAMs", 
                          "9" = "Proliferating Epithelial", 
                          "10" = "T cells", 
                          "11" = "Proliferating TAMs",
                          "12" = "TAMs", 
                          "13" = "Epithelial", 
                          "14" = "Epithelial", 
                          "15" = "Epithelial", 
                          "16" = "Endothelial", 
                          "17" = "Proliferating Epithelial", 
                          "18" = "Proliferating Fibroblast 1", 
                          "19" = "Epithelial", 
                          "20" = "Fibroblast 1", 
                          "21" = "Fibroblast 1", 
                          "22" = "B Cells", 
                          "23" = "Neutrophils",
                          "24" = "Neutrophils", 
                          "25" = "DCs", 
                          "26" = "Fibroblast 2", 
                          "27" = "RBC", 
                          "28" = "Acinar", 
                          "29" = "NK cells", 
                          "30" = "Proliferating T cells", 
                          "31" = "Unk Myeloid", 
                          "32" = "Myeloid & T cell mix", 
                          "33" = "TAMs", 
                          "34" = "Fibroblast 1",
                          "35" = "Endothelial", 
                          "36" = "RBC", 
                          "37" = "TAMs",
                          "38" = "Proliferating Fibroblast 1", 
                          "39" = "Duct", 
                          "40" = "TAMs",
                          "41" = "TAMs", 
                          "42" = "RBC", 
                          "43" = "Unk Myeloid",
                          "44" = "Endothelial", 
                          "45" = "Endothelial")
OrthotopicTCDD[["collapsed_clusters"]] <- OrthotopicTCDD@active.ident
DimPlot(object = OrthotopicTCDD_1, reduction = "umap", label = FALSE, pt.size = 1, split.by = "Group")
save(OrthotopicTCDD,file="OrthotopicTCDD.RData")

#For Main Fig UMAP
Idents(OrthotopicTCDD) <-"collapsed_clusters"
levels(OrthotopicTCDD)
OrthotopicTCDD_1 <-subset(OrthotopicTCDD, idents =  c("TAMs","Fibroblast 1","Neutrophils","Epithelial","Proliferating Epithelial","T cells","Proliferating TAMs","Endothelial","Proliferating Fibroblast 1","B Cells","DCs","Fibroblast 2","Acinar","NK cells","Proliferating T cells","Duct"))
order <- c("T cells","Fibroblast 1","Fibroblast 2","Neutrophils","TAMs","B Cells","DCs","NK cells","Duct","Epithelial","Endothelial","Acinar","Proliferating Epithelial","Proliferating Fibroblast 1","Proliferating T cells","Proliferating TAMs")
OrthotopicTCDD_1 @active.ident <- factor(OrthotopicTCDD_1 @active.ident, levels = order)
DimPlot(object = OrthotopicTCDD_1, reduction = "umap", label = F, pt.size = 0.5, cols = c("TAMs"= "gold",
                                                                                 "Fibroblast 1"="lightblue1",
                                                                                 "Neutrophils"="darkorange",
                                                                                 "Epithelial"="rosybrown2",
                                                                                 "Proliferating Epithelial"= "brown",
                                                                                 "T cells"= "royalblue",
                                                                                 "Proliferating TAMs"="indianred2",
                                                                                 "Endothelial"="mediumpurple",
                                                                                 "Proliferating Fibroblast 1"= "orchid1", 
                                                                                 "B Cells"="grey20",
                                                                                 "DCs"="green",
                                                                                 "Fibroblast 2"="plum",
                                                                                 "Acinar"="green4",
                                                                                 "NK cells"="gold3",
                                                                                 "Proliferating T cells"="cyan",
                                                                                 "Duct"="wheat"))


#To generate a Heatmap subset Cells
Idents(OrthotopicTCDD) <-"collapsed_clusters"
levels(OrthotopicTCDD)
Tcells <-subset(OrthotopicTCDD, idents =  c("T cells"))
Fibro <-subset(OrthotopicTCDD, idents =  c("Fibroblast 1", "Fibroblast 2"))
TAMs <-subset(OrthotopicTCDD, idents =  c("TAMs"))
Neutro <-subset(OrthotopicTCDD, idents =  c("Neutrophils"))

#Tcells Heatmap
Idents(Tcells) <-"Group"
levels(Tcells)       
Tcells.markers <- FindAllMarkers(Tcells, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Tcells.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- Tcells.markers$gene[grep("^ro|^mt-", Tcells.markers$gene)]
Tcells.markers<- Tcells.markers %>% filter(!gene %in% rp_mt_genes)
top10Tcells <- Tcells.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(Tcells, features = top10Tcells$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))
VlnPlot(object = Tcells, features = c("Foxp3","Tcf7"))

#Fibro Heatmap
Idents(Fibro) <-"Group"
levels(Fibro)       
Fibro.markers <- FindAllMarkers(Fibro, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Fibro.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- Fibro.markers$gene[grep("^ro|^mt-", Fibro.markers$gene)]
Fibro.markers<- Fibro.markers %>% filter(!gene %in% rp_mt_genes)
top10Fibro <- Fibro.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(Fibro, features = top10Fibro$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))

#TAMs Heatmap
Idents(TAMs) <-"Group"
levels(TAMs)       
TAMs.markers <- FindAllMarkers(TAMs, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
TAMs.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- TAMs.markers$gene[grep("^ro|^mt-", TAMs.markers$gene)]
TAMs.markers<- TAMs.markers %>% filter(!gene %in% rp_mt_genes)
top10TAMs <- TAMs.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(TAMs, features = top10TAMs$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))
VlnPlot(object = TAMs, features = c("Cxcl9","Tph1"))

#Neutro Heatmap
Idents(Neutro) <-"Group"
levels(Neutro)       
Neutro.markers <- FindAllMarkers(Neutro, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Neutro.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- Neutro.markers$gene[grep("^ro|^mt-", Neutro.markers$gene)]
Neutro.markers<- Neutro.markers %>% filter(!gene %in% rp_mt_genes)
top10Neutro <- Neutro.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(Neutro, features = top10Neutro$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))
VlnPlot(object = Neutro, features = c("Cxcl3"))

#Cell Prop
Idents(object = OrthotopicTCDD_1) <- 'collapsed_clusters'
levels(OrthotopicTCDD_1)

#splitting samples
samples.list <- unique(OrthotopicTCDD_1$Group)
clusters <- lapply(samples.list, function(x){
  subset <- subset(OrthotopicTCDD_1, subset = Group == x)
  dist <- data.frame(table(subset$collapsed_clusters))
  
  return(dist)
})

names(clusters) <- samples.list

#calculate relative freq (fractions) of each cell type
clusters_percent <- lapply(clusters, FUN = function(x){
  summ <- sum(x$Freq)
  x$Freq <- (x$Freq/summ)
  return(x)
})
Idents(object = OrthotopicTCDD_1) <- "Group"
levels(OrthotopicTCDD_1)
#making things ggplot-friendly!
clusters_dist <- reshape2::melt(clusters, id.var = "Var1")
colnames(clusters_dist) <- c("collapsed_clusters","variable","value","Group")
clusters_percent_dist <- reshape2::melt(clusters_percent, id.var = "Var1")
colnames(clusters_percent_dist) <- c("collapsed_clusters","variable","value","Group")


ggplot(clusters_percent_dist, aes(fill=collapsed_clusters, y = value, x = Group)) + 
  scale_x_discrete(limits = c("Vehicle","TCDD")) + 
  geom_bar(position = "stack", stat = "identity") + 
  theme_bw()+
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, size = 8, face = "bold", hjust = 1))+
  ylab("count") + xlab("Group") + ggtitle("Relative Cell Types Abundance") +
  scale_fill_manual(values = c("gold","lightblue1","darkorange","rosybrown2","brown", "royalblue","indianred2","mediumpurple","orchid1","grey20","green","plum","green4","gold3","cyan","wheat",'grey','grey','grey'))

DotPlot(OrthotopicTCDD_1, features = c("Cd3e","Pdgfra","Pdgfrb","Dcn","Col1a2","Itgam","S100a8","S100a9","Mmp9","C1qa","C1qb","Lyz2","Cd79a","Cd19","Itgax","Nkg7","Ceacam1","Krt19","Cdh1","Muc1","Msln","Vwf","Cdh5","Spink1","Amy1","Mki67"),cols = "RdBu", dot.scale = 8) + RotatedAxis()



######Subset fibroblasts
Idents(OrthotopicTCDD) <-"collapsed_clusters"
levels(OrthotopicTCDD)
Fibroblast<-subset(OrthotopicTCDD, idents =  c("Fibroblast 1", "Fibroblast 2"))

#Normalization ScaleData and UMAP Generation
Fibroblast <- NormalizeData(Fibroblast)
Fibroblast <- FindVariableFeatures(Fibroblast, selection.method = "vst", nfeatures = 2000)
Fibroblast <- ScaleData(Fibroblast, verbose = T, features = row.names(Fibroblast))
Fibroblast <- RunPCA(object = Fibroblast)
stdev <- Fibroblast@reductions$pca@stdev
var <- stdev^2
sum(var[1:32])/ sum(var) #include a number that leads to 0.90 var

PCNum = 32 #put the dimensions here

#Find Neighbors + Find Clusters (without harmony batch correction)
Fibroblast <- FindNeighbors(object = Fibroblast, dims = 1:PCNum)
Fibroblast <- FindClusters(object = Fibroblast, resolution = 0.8)

#Run UMAP and get unlabelled cluster UMAP and violin plot
Fibroblast<- RunUMAP(object = Fibroblast, dims = 1:PCNum)
DimPlot(object = Fibroblast, reduction = "umap", label = TRUE, pt.size = 1, split.by = 'Group')

#Feature plot fibrolabst subset
FeaturePlot(object = Fibroblast, features = c("Col1a2","Pdgfra","Pdgfrb", "Mki67"), cols = c("grey", "red"), order = TRUE, reduction = "umap", pt.size = .5)
FeaturePlot(object = Fibroblast, features = c("Clec3b","Il6","Col14a1", "Has1"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #iCAF
FeaturePlot(object = Fibroblast, features = c("Tagln", "Thy1", "Col12a1", "Thbs2", "Acta2"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #myCAF
FeaturePlot(object = Fibroblast, features = c("H2-Ab1", "Cd74", "Saa3", "Slpi"), cols = c("grey", "red"), reduction = "umap", pt.size = .5) #apCAF

VlnPlot(object = Fibroblast, features = c("Acta2"))
VlnPlot(object = Fibroblast, features = c("Tagln"))
VlnPlot(object = Fibroblast, features = c("Col12a1"))
VlnPlot(object = Fibroblast, features = c("Cd74"))

#Fibro Heatmap
Idents(Fibroblast) <- "seurat_clusters"
levels(Fibroblast)       
Fibroblast.markers <- FindAllMarkers(Fibroblast, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Fibroblast.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- Fibroblast.markers$gene[grep("^ro|^mt-", Fibroblast.markers$gene)]
Fibroblast.markers<- Fibroblast.markers %>% filter(!gene %in% rp_mt_genes)
top10Fibroblast <- Fibroblast.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(Fibroblast, features = top10Fibroblast$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))
VlnPlot(object = Fibroblast, features = c("Cxcl3"))

#Save fibroblast object
save(Fibroblast,file="Fibroblast.RData")

#Lable Clusters
Idents(Fibroblast) <- "Group"
new_order <- c("Vehicle", "TCDD")
Fibroblast@active.ident <- factor(Fibroblast@active.ident, levels = new_order)
Fibroblast[["Group"]] <- Fibroblast@active.ident 
Idents(Fibroblast) <- "seurat_clusters"
Fibroblast <- RenameIdents(Fibroblast,
                               "0" = "Fibro1", 
                               "1" = "Fibro2", 
                               "2" = "TCDD fibro", 
                               "3" = "TCDD fibro", 
                               "4" = "Fibro3", 
                               "5" = "Fibro4", 
                               "6" = "Fibro5", 
                               "7" = "Fibro6", 
                               "8" = "Fibro7", 
                               "9" = "Fibro8", 
                               "10" = "Proliferating fibro", 
                               "11" = "Fibro9",
                               "12" = "Fibro10", 
                               "13" = "Proliferating fibro", 
                               "14" = "Fibro11" )
Fibroblast[["collapsed_clusters"]] <- Fibroblast@active.ident
DimPlot(object = Fibroblast, reduction = "umap", label = FALSE, pt.size = 1, split.by = "Group")
save(Fibroblast,file="Fibroblast.RData")

#Rerun heatmap with collapsed fibro labels
Idents(Fibroblast) <- "collapsed_clusters"
levels(Fibroblast)       
Fibroblast.markers <- FindAllMarkers(Fibroblast, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Fibroblast.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
rp_mt_genes <- Fibroblast.markers$gene[grep("^ro|^mt-", Fibroblast.markers$gene)]
Fibroblast.markers<- Fibroblast.markers %>% filter(!gene %in% rp_mt_genes)
top10Fibroblast <- Fibroblast.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(Fibroblast, features = top10Fibroblast$gene) +  scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")))
VlnPlot(object = Fibroblast, features = c("Cyp1b1"), split.by = "Group", cols=c("lightblue", "blue"))
VlnPlot(object = Fibroblast, features = c("Cxcl14"), split.by = "Group", cols=c("lightblue", "blue"))
VlnPlot(object = Fibroblast, features = c("Mmp3"), split.by = "Group", cols=c("lightblue", "blue"))

#Find number of T cells
Idents(OrthotopicTCDD) <- "Group"
table(OrthotopicTCDD@active.ident)



```







